ergo-chain-generation/src/chain_generation.rs

```diff
@@ -0,0 +1,464 @@
+    } else {
+        // Autolykos v. 2: H(j|h|M) (line 5 from the Algo 2 of the spec)
+        let mut concat = vec![];
+        concat.extend(index_bytes);
+        concat.extend(height_bytes);
+        concat.extend(ergo_nipopow::test::calc_big_m());
+        BigInt::from_bytes_be(Sign::Plus, &blake2b256_hash(&concat).0[1..])
+    }
+}
+
+/// One way cryptographic hash function that produces numbers in [0,q) range.
+/// It calculates Blake2b256 hash of a provided input and checks whether the result is
+/// in range from 0 to a maximum number divisible by q without remainder.
+/// If yes, it returns the result mod q, otherwise make one more iteration using hash as an input.
+/// This is done to ensure uniform distribution of the resulting numbers.
+fn numeric_hash(input: &[u8], valid_range: BigInt, order: BigInt) -> BigInt {
+    let mut hashed: Vec<u8> = blake2b256_hash(input).into();
+    loop {
+        let bi = BigInt::from_bytes_be(Sign::Plus, &hashed);
+        if bi < valid_range {
+            break bi.modpow(&BigInt::from(1_u8), &order);
+        } else {
+            hashed = blake2b256_hash(&hashed).into();
+        }
+    }
+}
+
+/// Used in the miner when a BlockTransaction instance is not generated yet (because a header is not known)
+fn transactions_root(txs: &[Transaction], block_version: u8) -> Digest32 {
+    if block_version == 1 {
+        let tree = MerkleTreeNode::new(
+            txs.iter()
+                .map(|tx| {
+                    blake2b256_hash(&tx.bytes_to_sign().unwrap())
+                        .0
+                        .as_ref()
+                        .to_vec()
+                })
+                .collect(),
+        );
+        tree.hash().into()
+    } else {
+        let tree = MerkleTreeNode::new(
+            txs.iter()
+                .map(|tx| {
+                    let mut data = blake2b256_hash(&tx.bytes_to_sign().unwrap())
+                        .0
+                        .as_ref()
+                        .to_vec();
+                    //  Id of transaction "witness" (taken from Bitcoin jargon, means commitment to
+                    //  signatures of a transaction).  Id is 248-bit long, to distinguish
+                    //  transaction ids from witness ids in Merkle tree of transactions, where both
+                    //  kinds of ids are written into leafs of the tree.
+                    let witness: Vec<u8> = tx
+                        .inputs
+                        .iter()
+                        .flat_map(|input| {
+                            let i: Vec<u8> = input.spending_proof.proof.clone().into();
+                            i
+                        })
+                        .collect();
+                    data.extend(witness);
+                    data
+                })
+                .collect(),
+        );
+        tree.hash().into()
+    }
+}
+
+#[allow(clippy::unwrap_used)]
+#[cfg(test)]
+mod tests {
```

Comment:
These tests aren't found in the reference implementation, but I think it's worth running it through the more realistic chain-generator above.

Reply:
Great! The more tests the merrier! :) I appreciate it!